internal/runtimehandlerhooks/high_performance_hooks.go

```diff
@@ -356,6 +417,171 @@ func setCPUQuota(cpuMountPoint, parentDir string, c *oci.Container, enable bool)
 	return nil
 }
 
+// setCPUPMQOSResumeLatency sets the pm_qos_resume_latency_us for a cpu and stores the original
+// value so it can be restored later. If the latency is an empty string, the original latency
+// value is restored.
+func setCPUPMQOSResumeLatency(c *oci.Container, latency, cpuDir, cpuSaveDir string) error {
+	lspec := c.Spec().Linux
+	if lspec == nil ||
+		lspec.Resources == nil ||
+		lspec.Resources.CPU == nil ||
+		lspec.Resources.CPU.Cpus == "" {
+		return errors.Errorf("find container %s CPUs", c.ID())
+	}
+
+	cpus, err := cpuset.Parse(lspec.Resources.CPU.Cpus)
+	if err != nil {
+		return err
+	}
+
+	for _, cpu := range cpus.ToSlice() {
+		if latency != "" {
+			// retrieve the current latency
+			latencyFile := fmt.Sprintf("%s/cpu%d/power/pm_qos_resume_latency_us", cpuDir, cpu)
+			latencyOrig, err := ioutil.ReadFile(latencyFile)
+			if err != nil {
+				return err
+			}
+
+			// save the current latency so we can restore it later
+			cpuPowerSaveDir := fmt.Sprintf("%s/cpu%d/power", cpuSaveDir, cpu)
+			err = os.MkdirAll(cpuPowerSaveDir, 0o750)
+			if err != nil {
+				return err
+			}
+			latencyFileOrig := fmt.Sprintf("%s/pm_qos_resume_latency_us", cpuPowerSaveDir)
+			err = ioutil.WriteFile(latencyFileOrig, latencyOrig, 0o644)
+			if err != nil {
+				return err
+			}
+
+			// update the pm_qos_resume_latency_us
+			err = ioutil.WriteFile(latencyFile, []byte(latency), 0o644)
+			if err != nil {
+				return err
+			}
+		} else {
+			// retrieve the original latency
+			latencyFileOrig := fmt.Sprintf("%s/cpu%d/power/pm_qos_resume_latency_us", cpuSaveDir, cpu)
+			latencyOrig, err := ioutil.ReadFile(latencyFileOrig)
+			if err != nil {
+				if errors.Is(err, os.ErrNotExist) {
+					// the latency may have already been restored by a previous invocation of the hook
+					return nil
+				}
+				return err
+			}
+
+			// restore the original latency
+			latencyFile := fmt.Sprintf("%s/cpu%d/power/pm_qos_resume_latency_us", cpuDir, cpu)
+			err = ioutil.WriteFile(latencyFile, latencyOrig, 0o644)
+			if err != nil {
+				return err
+			}
+
+			// remove the saved latency
+			err = os.Remove(latencyFileOrig)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// setCPUFreqGovernor sets the scaling_governor for a cpu and stores the original
+// value so it can be restored later. If the governor is an empty string, the original
+// scaling_governor value is restored.
+func setCPUFreqGovernor(c *oci.Container, governor, cpuDir, cpuSaveDir string) error {
```

Comment:
Since `cpuDir` and `cpuSaveDir` never change, it does not make sense to have them as parameters.

For the test case, you can maybe do `chroot`, or make `sysCPUDir` and `sysCPUSaveDir` variables that are changed by the test case.

Reply:
I based this code on the existing code in this file where the paths are passed in as parameter in order to facilitate unit testing. When I wrote the code, I also thought this was strange so I did some research to understand if there was a nice way for unit tests to mock out calls to os.ReadFile, os.WriteFile, etc... In short, there does not seem to be a reasonable way to do this. The best solution I found (other than passing in the paths which is what most people do) is to pass in a filesystem instead of the paths, which allows the unit test to pass in a "fake" filesystem. This would be a variation on https://talks.golang.org/2012/10things.slide#8

I don't see any testcases in crio (or elsewhere) that use chroot for this. I also don't want to make sysCPUDir and sysCPUSaveDir into global variables to facilitate unit testing - that also has potential issues (e.g. with parallel testcase execution).

So... given that I'm following an existing pattern from code in this file and there don't seem to be any existing examples of this type of mocking being done in the crio-o codebase (correct me if I'm wrong), I'd prefer to leave this as is.

Reply:
we can make a note above the functions that say this reasoning for keeping them parameters. If this is the way we're doing it now, we can leave it and possibly address in the future

Reply:
One other way to solve this is to have something like
```
func setCPUFreqGovernor(c *oci.Container, governor, cpuDir, cpuSaveDir string) error {
   return doSetCPUFreqGovernor(c, governor, sysCPUDir, sysCPUSaveDir)
}

func doSetCPUFreqGovernor(c *oci.Container, governor, cpuDir, cpuSaveDir string) error {
....
}
```

then use the first function in the code, and the second function in the tests.

Overall it's not a big deal, so it's fine the way it is, but please consider refactoring the code later to simplify it.

Reply:
I like that - it keeps the calling functions simple (no need to pass the paths) but allows for clean and simple unit tests. I made the change.